### PR TITLE
fix: close GRPC client connection only when all plugin connections are disabled

### PR DESCRIPTION
**Pull request by vinay-gopalan**
This PR removes the closing the GRPC client connection from the GRPC Backend Client. Previously we were prematurely closing a plugin connection which was causing multiplexed plugins to break. Now, the responsibility of closing the GRPC Client connection solely rests with `go-plugin` and its `Kill` function. This also resolves a pre-existing bug in Vault wherein all plugin processes were failing to exit gracefully and were having to be killed forcefully.

### CODE REVIEW

File: vault/plugin_catalog.go [hunk 1]
```diff
@@ -214,6 +214,13 @@ func (c *PluginCatalog) cleanupExternalPlugin(name, id string) error {
 		}
 		c.logger.Debug("killed external plugin process", "name", name, "pid", pc.pid)
 	} else if len(extPlugin.connections) == 0 || pc.client.Exited() {
```
Comment (calvn) [inline vault/plugin_catalog.go:216]: Do we also want to close the connection right before L210 for the non-multiplexing case?

Comment (vinay-gopalan) [reply vault/plugin_catalog.go:222]: Yup! You were one step ahead of me 😅 Fixed now

File: vault/plugin_catalog.go [hunk 2]
```diff
@@ -207,6 +207,12 @@ func (c *PluginCatalog) cleanupExternalPlugin(name, id string) error {
 	c.logger.Debug("removed plugin client connection", "id", id)
 
 	if !extPlugin.multiplexingSupport {
+		clientConn, err := pc.client.Client()
+		if err != nil {
+			return err
```
Comment (calvn) [inline vault/plugin_catalog.go:212]: We might want to make this best-effort and simply log the error, but continue with the rest of the cleanup.

File: sdk/plugin/grpc_backend_client.go [hunk 3]
```diff
@@ -194,7 +194,6 @@ func (b *backendGRPCPluginClient) Cleanup(ctx context.Context) {
 	if server != nil {
 		server.(*grpc.Server).GracefulStop()
 	}
-	b.clientConn.Close()
```
Comment (briankassouf) [inline sdk/plugin/grpc_backend_client.go:197]: If this is the only reason we are stashing the connection as a member variable can we delete the variable from the backendGRPCPluginClient struct?

Comment (vinay-gopalan) [reply sdk/plugin/grpc_backend_client.go:197]: Good call, removed in d2b0eb95a9c2c40b5590818a2f62c455624f33aa

Comment (austingebauer) [reply sdk/plugin/grpc_backend_client.go:197]: I think you'll need to remove in https://github.com/hashicorp/vault/blob/main/sdk/plugin/backend.go#L62 or it won't build.

Comment (vinay-gopalan) [reply sdk/plugin/grpc_backend_client.go:197]: Whoops! Thanks 🙇🏼

File: builtin/logical/aws/iam_policies_test.go [hunk 4]
```diff
@@ -207,7 +207,7 @@ func Test_combinePolicyDocuments(t *testing.T) {
 				`{"Version": "2012-10-17", "Statement": [{"Effect": "Allow", "NotAction": "ec2:DescribeAvailabilityZones", "Resource": "*"}]}`,
 			},
 			expectedOutput: `{"Version": "2012-10-17","Statement":[{"Effect": "Allow","NotAction": "ec2:DescribeAvailabilityZones",	"Resource": "*"}]}`,
-			expectedErr:    false,
+			expectedErr: false,
```
Comment (austingebauer) [inline builtin/logical/aws/iam_policies_test.go:210]: Hmm, any idea where this came from? Vault main isn't failing the formatting check, so you might want to pull latest from there.

Comment (vinay-gopalan) [reply builtin/logical/aws/iam_policies_test.go:210]: This actually happened after I ran `make fmt`. The `fmt` check in CI was failing, and so I ran `make fmt` hoping to resolve it... But looks like the `fmt` check is still failing 🤔

Comment (austingebauer) [reply builtin/logical/aws/iam_policies_test.go:210]: Huh, I just pull the latest from your branch here and `make fmt` results in a new diff. I think try `make fmt` again and commit the diff should work.

Comment (vinay-gopalan) [reply builtin/logical/aws/iam_policies_test.go]: Fixed in latest commit, thanks!

**Pull request merged**